OpenChange/MAPIStoreMailMessage.m

```diff
@@ -396,36 +401,82 @@ - (void) _fetchBodyData
-              NSString *charset = [bodyPartsCharsets objectForKey: key];
+              charset = [bodyPartsCharsets objectForKey: key];
               if (charset)
                 {
-                  NSString *stringValue = [content bodyStringFromCharset: charset];
-                  [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                  NSString *stringValue = nil;
+                  if (html)
+                    {
+                      if ([charset isEqualToString: headerCharset])
+                        {
+                          [htmlContent appendData: content];
+                        }
+                      else
+                        {
+                          stringValue = [content bodyStringFromCharset: charset];
+                          [htmlContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                        }
+                      if (multipartMixed)
+                        {
+                          if (stringValue == nil)
+                            stringValue = [content bodyStringFromCharset: charset];
+                          [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                        }
+                    }
+                  else
+                    {
+                      /* No HTML, then is text/plain */
+                      stringValue = [content bodyStringFromCharset: charset];
+                      [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                      if (multipartMixed)
+                        {
+                          stringValue = [stringValue stringByReplacingOccurrencesOfString: @"\n"
+                                                                               withString: @"<br/>"];
+                          [htmlContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                        }
+                    }
                 }
               else
                 {
-                  [textContent appendData: content];
+                  /* Without charset we cannot mangle the text, so we add as it stands */
+                  if (html || multipartMixed)
+                    [htmlContent appendData: content];
+                  if (!html || multipartMixed)
+                    [textContent appendData: content];
                 }
+
             }
-          else if (mailIsEvent &&
-                   ([mimeType isEqualToString: @"text/calendar"] ||
-                    [mimeType isEqualToString: @"application/ics"]))
+          else if ([mimeType isEqualToString: @"text/calendar"] ||
+                   [mimeType isEqualToString: @"application/ics"])
             {
-              content = [content bodyDataFromEncoding: encoding];
               [textContent appendData: content];
+              [htmlContent appendData: content];
```

Comment:
It is to preserve the old behaviour. In the old behaviour in mixed and calendar messsages textContent and htmlContent pointed to the same NSMutableData. However since now they can contain different data  even when mixed(<br/> and it could be mixed and alternatives parts in the same message (#247)  they must be always different NSMutableData.